src/coord/src/session/vars.rs

```diff
@@ -773,6 +794,139 @@ impl Value for i32 {
     }
 }
 
+const SEC_TO_MIN: u64 = 60u64;
+const SEC_TO_HOUR: u64 = 60u64 * 60;
+const SEC_TO_DAY: u64 = 60u64 * 60 * 24;
+const MICRO_TO_MILLI: u32 = 1000u32;
+
+impl Value for Duration {
+    const TYPE_NAME: &'static str = "duration";
+
+    fn parse(s: &str) -> Result<Duration, ()> {
+        let s = s.trim();
+        // Take all numeric values from [0..]
+        let split_pos = s
+            .chars()
+            .position(|p| !char::is_numeric(p))
+            .unwrap_or_else(|| s.chars().count());
+
+        // Error if the numeric values don't parse, i.e. there aren't any.
+        let d = s[..split_pos].parse::<u64>().map_err(|_| ())?;
+
+        // We've already trimmed end
+        let (f, m): (fn(u64) -> Duration, u64) = match s[split_pos..].trim_start() {
+            "us" => (Duration::from_micros, 1),
+            // Default unit is milliseconds
+            "ms" | "" => (Duration::from_millis, 1),
+            "s" => (Duration::from_secs, 1),
+            "min" => (Duration::from_secs, SEC_TO_MIN),
+            "h" => (Duration::from_secs, SEC_TO_HOUR),
+            "d" => (Duration::from_secs, SEC_TO_DAY),
+            _ => return Err(()),
+        };
+
+        let d = if d == 0 {
+            Duration::from_secs(u64::MAX)
+        } else {
+            f(d.checked_mul(m).ok_or(())?)
+        };
+        Ok(d)
+    }
+
+    // The strategy for formatting these strings is to find the least
+    // significant unit of time that can be printed as an integer––we know this
+    // is always possible because the input can only be an integer of a single
+    // unit of time.
+    fn format(&self) -> String {
+        let micros = self.subsec_micros();
+        if micros > 0 {
+            match micros {
+                ms if ms != 0 && ms % MICRO_TO_MILLI == 0 => {
+                    format!(
+                        "{} ms",
+                        self.as_secs() * 1000 + u64::from(ms / MICRO_TO_MILLI)
+                    )
+                }
+                us => format!("{} us", self.as_secs() * 1_000_000 + u64::from(us)),
+            }
+        } else {
+            match self.as_secs() {
+                zero if zero == u64::MAX => "0".to_string(),
```

Comment:
I feel like this should still get a unit? Or is 0 a special duration in postgres that we're emulating?